src/go/rpk/pkg/serde/proto.go

```diff
@@ -0,0 +1,123 @@
+// returns the record encoded in the protobuf wire format.
+func EncodeProtoRecord(ctx context.Context, cl *sr.Client, schema *sr.Schema, protoFQN string, schemaID int, record []byte) ([]byte, error) {
+	if protoFQN == "" {
+		return nil, fmt.Errorf("please provide the protobuf message name")
+	}
+	// Compile the Schema Registry proto. //
+	const inMemFileName = "tmp.proto"
+	accessorMap := make(map[string]string)
+	// This is the original schema.
+	accessorMap[inMemFileName] = schema.Schema
+
+	// And we add the rest of schemas if we have references.
+	if len(schema.References) > 0 {
+		err := mapReferences(ctx, cl, schema, accessorMap)
+		if err != nil {
+			return nil, fmt.Errorf("unable to map references: %v", err)
+		}
+	}
+
+	compiler := protocompile.Compiler{
+		Resolver: &protocompile.SourceResolver{
+			Accessor: protocompile.SourceAccessorFromMap(accessorMap),
+		},
+		SourceInfoMode: protocompile.SourceInfoStandard,
+	}
+	compiled, err := compiler.Compile(ctx, inMemFileName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compile the given schema: %v", err)
+	}
+
+	// Create proto message with runtime information. //
+	d := compiled.FindFileByPath(inMemFileName).FindDescriptorByName(protoreflect.FullName(protoFQN))
+	if d == nil {
+		return nil, fmt.Errorf("unable to find message with name %q", protoFQN)
+	}
+	msgDescriptor, ok := d.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("unable to process message with name %q", protoFQN)
+	}
+	message := dynamicpb.NewMessage(msgDescriptor)
+
+	// Unmarshal the record into the proto message. //
+	o := protojson.UnmarshalOptions{
+		Resolver: compiled.AsResolver(),
+	}
+	err = o.Unmarshal(record, message)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode the record into the given schema: %v", err)
+	}
+
+	// Encode the proto message to protobuf wire format. //
+	binary, err := proto.Marshal(message)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encode the record: %v", err)
+	}
+
+	// Add schema registry wire format header. //
+	var serdeHeader sr.ConfluentHeader
+	h, _ := serdeHeader.AppendEncode(nil, schemaID, messageIndex(msgDescriptor))
+	return append(h, binary...), nil
+}
+
+// messageIndex traverse the message descriptor to get the list of
+// message-indexes which is needed in the wire format.
+func messageIndex(d protoreflect.MessageDescriptor) []int {
+	idx := []int{d.Index()}
+	if parent := d.Parent(); parent != nil {
+		pmsg, ok := parent.(protoreflect.MessageDescriptor)
+		// If !ok, we have reached the top, and it's a FileDescriptor.
+		if !ok {
+			return idx
+		}
+		return append(messageIndex(pmsg), idx...)
+	}
+	return idx
+}
+
+// mapReferences recursively inserts the schema references into the given
+// refMap. It uses the client to retrieve the schema references.
+func mapReferences(ctx context.Context, cl *sr.Client, schema *sr.Schema, refMap map[string]string) error {
```

Comment:
I think it should be a separate step to resolve references. It took me a second read to see that parseReferences only runs once, because references are recursively resolved only once.

Rather than resolving references at the time we go to encode a record, I think we should first separate fully recursively parsing sr.Schema, and _then_ drop into record encoding.

I think this could be done with something like,

```go
func ResolveReferences(ctx context.Context, cl *sr.Client, schema *sr.Schema)
```

and then you could drop ctx and cl from Encode.